arduino-ide-extension/src/common/protocol/core-service.ts

```diff
@@ -9,6 +17,78 @@ export const CompilerWarningLiterals = [
   'All',
 ] as const;
 export type CompilerWarnings = typeof CompilerWarningLiterals[number];
+export namespace CoreError {
+  export type ErrorInfo = CliErrorInfo;
+  export interface Compiler extends ErrorInfo {
+    readonly message: string;
+    readonly location: Location;
+  }
+  export namespace Compiler {
+    export function is(error: ErrorInfo): error is Compiler {
+      const { message, location } = error;
+      return !!message && !!location;
+    }
+    export function toString({ location, message }: Compiler): string {
+      const {
+        uri,
+        range: {
+          start: { line, character },
+        },
+      } = location;
+      return `${new URI(uri)
+        .withFragment(`L${line},${character}`)
+        .toString()} => '${message}'`;
+    }
+    export function sameAs(left: Compiler, right: Compiler): boolean {
+      const sameRange = (left: Range, right: Range) =>
+        left.start.line === right.start.line &&
+        left.start.character === right.start.character &&
+        left.end.line === right.end.line &&
+        left.end.character === right.end.character;
+      const sameLocation = (left: Location, right: Location) =>
+        left.uri === right.uri && sameRange(left.range, right.range);
+      return (
+        left.message === right.message &&
+        left.details === right.details &&
+        sameLocation(left.location, right.location)
+      );
+    }
+  }
+  export const Codes = {
+    Verify: 4001,
+    Upload: 4002,
+    UploadUsingProgrammer: 4003,
+    BurnBootloader: 4004,
+  };
+  export const VerifyFailed = create(Codes.Verify);
+  export const UploadFailed = create(Codes.Upload);
+  export const UploadUsingProgrammerFailed = create(
+    Codes.UploadUsingProgrammer
+  );
+  export const BurnBootloaderFailed = create(Codes.BurnBootloader);
+  export function is(
+    error: unknown
+  ): error is ApplicationError<number, ErrorInfo[]> {
+    return (
+      error instanceof Error &&
+      ApplicationError.is(error) &&
+      Object.values(Codes).includes(error.code)
+    );
+  }
+  function create(
+    code: number
+  ): ApplicationError.Constructor<number, ErrorInfo[]> {
+    return ApplicationError.declare(
+      code,
+      (message: string, data: ErrorInfo[]) => {
```

Comment:
Nice to have: consider replacing `message` with a `ServiceError` instance.

Reply:
The `ServiceError` object does not contain sufficient info about the error, so this won't fix.

Ref: https://github.com/arduino/arduino-cli/issues/1767